src/NuGet.Core/NuGet.Packaging/Signing/SignPackageArchive.cs

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.IO;
+using System.IO.Compression;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NuGet.Packaging.Signing
+{
+    /// <summary>
+    /// A nupkg that supports both reading and writing signatures.
+    /// </summary>
+    public class SignPackageArchive : PackageArchiveReader, ISignPackage
+    {
+        // TEMP
+        private const string TestSignedPath = "testsigned/signed.json";
+
+        public SignPackageArchive(ZipArchive zip)
+            : base(zip)
+        {
+        }
+
+        /// <summary>
+        /// Add a file to the package.
+        /// </summary>
+        public async Task AddAsync(string path, Stream stream, CancellationToken token)
+        {
+            var entry = Zip.CreateEntry(path, CompressionLevel.Optimal);
+
+            using (var writer = new StreamWriter(entry.Open()))
+            {
+                await stream.CopyToAsync(stream);
```

Comment:
This does not look right. You can copy file stream into another file stream and should not need a stream writer. Consider something like - 

```
                var entry = zip.CreateEntry(path, CompressionLevel.Optimal);
                using (Stream zipEntryStream = entry.Open())
                {
                    //Copy the file stream to the zip entry stream
                    fileStream.CopyToAsync(zipEntryStream);
                }
```

Not relevant just yet, but did not want this staying as is. :)